wasm: Add wasm support

tmc opened the pull request:
Add WebAssembly (WASM) Support to bbolt

This PR adds comprehensive support for WebAssembly targets (js and wasip1) to bolt, enabling the database to run in browsers and WASI-compliant runtimes.

What's Changed

- Memory-mapped I/O emulation: Since WASM doesn't support real mmap, we implement a buffer-based approach that reads the database file into memory
- Transaction synchronization: Added a txIniter interface to handle platform-specific transaction initialization, ensuring data consistency in WASM environments where memory buffers
need manual refresh
- Platform-specific implementations: Created bolt_wasm.go with WASM-specific implementations of mmap, munmap, madvise, and file locking operations

Testing Infrastructure

- Added WASM test targets in the Makefile:
  - make test-wasm for browser-based JavaScript WASM (GOOS=js)
  - make test-wasip1 for WASI Preview 1 runtime testing
- Added GitHub Actions workflows for automated WASM testing in CI

Test Adaptations
- Skip memory-intensive tests on WASM platforms due to runtime constraints
- Added appropriate test guards for WASM-specific limitations

Testing

All tests pass on WASM platforms:
- Browser WASM (GOOS=js GOARCH=wasm)
- WASI Preview 1 (GOOS=wasip1 GOARCH=wasm)

The implementation has been tested with both wazero and wasmtime runtimes.

tmc committed:
wasm: add WebAssembly (WASM) support

Add support for running bbolt in WebAssembly environments (js/wasm and wasip1):
- Implement memory management for WASM platforms without mmap
- Add platform-specific transaction initialization
- Configure smaller page size and memory limits for WASM
- Add GitHub workflow for WASM testing (js and wasip1)
- Adjust tests to handle WASM memory constraints
- Support both browser (js) and WASI environments

Signed-off-by: Travis Cline <travis.cline@gmail.com>

ahrtr commented:
cc @Elbehery  @fuweid @ivanvc @jmhbnz @tjungblu

tmc committed:
wasm: simplify platform-specific page size handling

Replace os.Getpagesize() calls with common.DefaultPageSize constant and remove redundant pagesize.go implementation. Simplify WASM-specific code by removing custom initialization and page size handling. Update tests to use the new constant.

Signed-off-by: Travis Cline <travis.cline@gmail.com>

tmc commented on internal/common/bolt_wasm.go:
```diff
@@ -0,0 +1,24 @@
+// On WASM platforms, we always return 4KB as the page size.
+func GetPagesize() int {
+	return 4096
+}
```
No, it returns 64k and I was forcing to work around constraints -- I'm pushing a simplification which pulls the wasm-specific override.

ivanvc commented:
I suggest holding this pull request until we make #986 work and find a solution to cross-compile in other architectures, so we're able to catch issues in new ones early/automated, rather than waiting for someone to find out compile/build issues.

Elbehery commented:
> I suggest holding this pull request until we make #986 work and find a solution to cross-compile in other architectures, so we're able to catch issues in new ones early/automated, rather than waiting for someone to find out compile/build issues.

cc @ahrtr

Elbehery commented on db.go:
```diff
@@ -772,12 +772,28 @@ func (db *DB) Logger() Logger {
 
+	// Allow WASM-specific transaction initialization
+	if runtime.GOARCH == "wasm" {
+		if initer, ok := any(db).(txIniter); ok {
```
ping @tmc

tmc replied:
type assertions must take an interface value and db is *DB which means it can't be used in a type assertion.

Elbehery replied:
cc @fuweid @ahrtr

tmc committed:
wasm: update db_test.go for WASM support changes

Update the db_test.go file to accommodate the recent WASM-related changes, including simplifying platform-specific page size handling and adjusting tests to handle WASM memory constraints.

Signed-off-by: Travis Cline <travis.cline@gmail.com>

Elbehery commented on db.go:
```diff
@@ -834,6 +850,16 @@ func (db *DB) beginRWTx() (*Tx, error) {
+	if runtime.GOARCH == "wasm" {
+		if initer, ok := any(db).(txIniter); ok {
+			if err := initer.txInit(); err != nil {
+				db.rwlock.Unlock()
```
ditto

Elbehery reviewed (commented):
LGTM 

the implementation of `bolt_wasm.go` is straight forward. 

But I am in favor of using `defer` for releasing the lock

cc @ahrtr @ivanvc
